### PR TITLE
Feat: add staking soft rate limit

### PR DESCRIPTION
**Pull request by TheDZhon**
Implement a soft staking rate limit.

### CODE REVIEW

**Comment by skozin** [comment]
I'd implement it a little bit differently, smth like this:

```solidity
contract LidoPseudoCode {
    /**
     * Staking is limited in rate by imposing limit on the stake amount
     * at each moment in time.
     *
     * ▲ Stake limit
     * │.....   .....    ........ ...            ....     ... Stake limit = max
     * │      .        .         .   .   .      .    . . .
     * │     .        .               . .  . . .      . .
     * │             .                 .  . . .               
     * │────────────────────────────────────────────────────> Time
     * │     ^       ^           ^   ^^^  ^ ^ ^     ^^^ ^     Stake events
     */
    struct StakeLimitState {
        // moment of the last stake event
        uint256 prevStakeTime;
        // stake limit after the last stake event
        uint256 prevStakeLimit;
        // the rate of stake limit increase per second
        uint256 stakeLimitIncPerSec;
        // max stake limit
        uint256 maxStakeLimit;
    }

    function stake(uint256 amount) external {
        StakeLimitState memory limitState = _readStateLimit();
        uint256 stakeLimit = _getStakeLimit(limitState);
        require(amount <= stakeLimit, "RATE_LIMIT_EXCEEDED");

        limitState.prevStakeTime = _getTime();
        limitState.prevStakeLimit = stakeLimit - amount;
        _writeStateLimit(limitState);

        _stake(amount);
    }

    function _getStakeLimit(StakeLimitState memory lim) internal pure returns (uint256) {
        uint256 dtSec = _getTime() - lim.prevStakeTime;
        return min(lim.maxStakeLimit, lim.prevStakeLimit + dtSec * lim.stakeLimitIncPerSec);
    }

    function min(uint256 a, uint256 b) internal pure returns (uint256) {
        return a < b ? a : b;
    }

    function _getTime() internal view returns (uint256) {
        return block.timestamp;
    }
}
```

**Comment by TheDZhon** [comment]
> I'd implement it a little bit differently, smth like this (pseudo-code):
> 
> ```solidity
> contract LidoPseudoCode {
>     /**
>      * Staking is rate-limited by imposing a limit on the stake amount
>      * at each moment in time.
>      *
>      * ▲ Stake limit
>      * │.....  .....   ........ ...            ....     ... Stake limit = max
>      * │      .       .        .   .   .      .    . . .
>      * │     .       .              . .  . . .      . .
>      * │            .                .  . . .               
>      * │──────────────────────────────────────────────────> Time
>      * │     ^      ^          ^   ^^^  ^ ^ ^     ^^^ ^     Stake events
>      */
>     struct StakeLimitState {
>         // moment of the last stake event
>         uint256 prevStakeTime;
>         // stake limit after the last stake event
>         uint256 prevStakeLimit;
>         // the rate of stake limit increase per second
>         uint256 stakeLimitIncPerSec;
>         // max stake limit
>         uint256 maxStakeLimit;
>     }
> 
>     function submit(uint256 amount) external {
>         StakeLimitState memory limitState = _readStakeLimitState();
>         uint256 stakeLimit = _getStakeLimit(limitState);
>         require(amount <= stakeLimit, "RATE_LIMIT_EXCEEDED");
> 
>         limitState.prevStakeTime = _getTime();
>         limitState.prevStakeLimit = stakeLimit - amount;
>         _writeStakeLimitState(limitState);
> 
>         _submit(amount);
>     }
> 
>     function _getStakeLimit(StakeLimitState memory lim) internal pure returns (uint256) {
>         uint256 dtSec = _getTime() - lim.prevStakeTime;
>         return min(lim.maxStakeLimit, lim.prevStakeLimit + dtSec * lim.stakeLimitIncPerSec);
>     }
> 
>     function min(uint256 a, uint256 b) internal pure returns (uint256) {
>         return a < b ? a : b;
>     }
> 
>     function _getTime() internal view returns (uint256) {
>         return block.timestamp;
>     }
> }
> ```

Thank you for the proposal!
If I understand correctly, you chose to change a period duration to the corresponding rate (which is `maxStakeLimit / stakeLimitPeriod`). This would work and may look more comprehensible. I suggest making a decision based on some other opinions.

**Comment by ujenjt** [comment]
@skozin @TheDZhon I really like how Sam approach this. Additionally I propose to implement public view for `getStakeLimit`, this would help frontends to build better UX. Also, we can extract packing/unpacking limit slot to the library.

**Comment by TheDZhon** [comment]
Ok, will upgrade the impl in the coming hours.
@skozin, @ujenjt thank you for the feedback :+1:

**Comment by ujenjt** [comment]
Regarding the rate vs period duration + limit, it's matter of taste, but I think that rate is more elegant and simple from the code readability perspective. But we can keep duration + limit if you see any additional advantages.

**Comment by folkyatina** [comment]
+1, Sam's variant is without division, so it's easier to understand and to avoid catch-ups.

I'm wondering, if we can use just block numbers instead of timestamp for limiting. It will make it even more resilient and simple, imo. Does the limit per sec is really better than limit per block?

**Comment by TheDZhon** [comment]
Finally, here we are.
image

I have refactored the code:
- new `StakeLimitUtils` library was added that hides bit-twiddling hacks;
- auth modifiers in `Lido.sol` changed to the trickery `_auth` helper which hugely lower bytecode's size :exploding_head:;
- staking rate-limit sets in terms of `maxStakeLimit`` and `stakeLimitIncreasePerBlock`;
- new external viewers to be potentially used by a frontend;
- test suites are enlarged and enriched.

@ujenjt, @skozin, @folkyatina please have a look if you can :pray:

**Review by folkyatina — COMMENTED** [review]
Checked it briefly. Will do another run in the morning.

File: contracts/0.4.24/Lido.sol [hunk 1]
```diff
@@ -120,19 +124,118 @@ contract Lido is ILido, IsContract, StETH, AragonApp {
...
+        uint96 _maxStakeLimit,
+        uint96 _stakeLimitIncreasePerBlock
+    ) external {
+        _auth(STAKING_RESUME_ROLE);
+
+        require(_maxStakeLimit >= _stakeLimitIncreasePerBlock, "TO_LARGE_INCREESE");
+        require(
+            (_stakeLimitIncreasePerBlock == 0)
+            || (_maxStakeLimit / _stakeLimitIncreasePerBlock <= uint32(-1)),
+            "TO_SMALL_INCREASE"
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply contracts/0.4.24/Lido.sol:187]: Yep, you are right!

File: contracts/0.4.24/Lido.sol [hunk 2]
```diff
@@ -120,19 +124,118 @@ contract Lido is ILido, IsContract, StETH, AragonApp {
...
+    * @param _maxStakeLimit max stake limit value
+    * @param _stakeLimitIncreasePerBlock stake limit increase per single block
+    */
+    function resumeStaking(
+        uint96 _maxStakeLimit,
+        uint96 _stakeLimitIncreasePerBlock
+    ) external {
+        _auth(STAKING_RESUME_ROLE);
+
+        require(_maxStakeLimit >= _stakeLimitIncreasePerBlock, "TO_LARGE_INCREESE");
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply contracts/0.4.24/Lido.sol:183]: On my way

File: contracts/0.4.24/Lido.sol [hunk 3]
```diff
@@ -120,19 +124,118 @@ contract Lido is ILido, IsContract, StETH, AragonApp {
...
+    *
+    * To disable rate-limit pass zero arg values.
+    * @dev Reverts if:
+    * - `_maxStakeLimit` < `_stakeLimitIncreasePerBlock`
+    * - `_maxStakeLimit` / `_stakeLimitIncreasePerBlock` >= 2^32 (only if `_stakeLimitIncreasePerBlock` != 0)
+    * Emits `StakeResumed` event
+    * @param _maxStakeLimit max stake limit value
+    * @param _stakeLimitIncreasePerBlock stake limit increase per single block
+    */
+    function resumeStaking(
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply contracts/0.4.24/Lido.sol:177]: No, you can't put staking on pause by calling `resumeStaking`

File: contracts/0.4.24/Lido.sol [hunk 4]
```diff
@@ -559,20 +664,34 @@ contract Lido is ILido, IsContract, StETH, AragonApp {
     * @return amount of StETH shares generated
     */
     function _submit(address _referral) internal whenNotStopped returns (uint256) {
-        address sender = msg.sender;
-        uint256 deposit = msg.value;
-        require(deposit != 0, "ZERO_DEPOSIT");
+        uint256 stakeLimitSlotValue = STAKE_LIMIT_POSITION.getStorageUint256();
+
+        require(!stakeLimitSlotValue.isStakingPaused(), "STAKING_PAUSED");
+        require(msg.value != 0, "ZERO_DEPOSIT");
```
Comment (arwer13) [inline contracts/0.4.24/Lido.sol:670]: Maybe move this to the top to fails earlier in the case?

File: contracts/0.4.24/lib/StakeLimitUtils.sol [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+// As a result, slot's memory aligned as follows:
+//
+// LSB ------------------------------------------------------------------------------> MSB
+// 0______________________32______________128_________________________160______________256
+// |______________________|________________|___________________________|________________|
+// | prevStakeBlockNumber | prevStakeLimit | maxStakeLimitGrowthBlocks | maxStakeLimit  |
+// |<----- 32 bits ------>|<-- 96 bits --->|<---------- 32 bits ------>|<--- 96 bits -->|
+//
+//
+// NB: we represent `maxStakeLimitGrowthPeriod` as follows:
```
Comment (arwer13) [inline contracts/0.4.24/lib/StakeLimitUtils.sol:21]: Is this comment about `maxStakeLimitGrowthPeriod` is obsolete?

File: contracts/0.4.24/Lido.sol (hunk 4 above)
Comment (TheDZhon) [reply contracts/0.4.24/Lido.sol:670]: Done, thx!

File: contracts/0.4.24/lib/StakeLimitUtils.sol (hunk 5 above)
Comment (TheDZhon) [reply contracts/0.4.24/lib/StakeLimitUtils.sol:21]: yep, that was a typo, ty!

File: contracts/0.4.24/lib/StakeLimitUtils.sol [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+        if (_stakeLimitIncPerBlock > 0) {
+            ret |= uint256(uint32(_maxStakeLimit / _stakeLimitIncPerBlock)) << MAX_STAKE_LIMIT_GROWTH_BLOCKS_OFFSET;
+        }
+    }
+
+    /**
+    * @notice Calculate stake limit for the current block.
+    */
+    function getCurrentStakeLimit(uint256 _slotValue) internal view returns(uint256 limit) {
```
Comment (folkyatina) [inline contracts/0.4.24/lib/StakeLimitUtils.sol:72]: Don't you want to call it `calculateCurrentStakeLimit`? It looks like the plain getter now, but it's tricky inside.

**Review by folkyatina — APPROVED** [review]
Didn't find anything suspicious. Minor nonblocking notes on naming.

File: contracts/0.4.24/lib/StakeLimitUtils.sol (hunk 6 above)
Comment (TheDZhon) [reply contracts/0.4.24/lib/StakeLimitUtils.sol:72]: Hmm, might work! Thank you, gonna change it.

**Pull request merged**